src/Microsoft.TestPlatform.CrossPlatEngine/Client/InProcessProxyexecutionManager.cs

```diff
@@ -0,0 +1,126 @@
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Engine;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Engine.ClientProtocol;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Engine.TesthostProtocol;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
+
+    internal class InProcessProxyExecutionManager : IProxyExecutionManager
+    {
+        private ITestHostManagerFactory testHostManagerFactory;
+        IExecutionManager executionManager;
+        public bool IsInitialized { get; private set; } = false;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="InProcessProxyexecutionManager"/> class.
+        /// </summary>
+        public InProcessProxyExecutionManager() : this(new TestHostManagerFactory())
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="InProcessProxyexecutionManager"/> class.
+        /// </summary>
+        /// <param name="testHostManagerFactory">
+        /// Manager factory
+        /// </param>
+        internal InProcessProxyExecutionManager(ITestHostManagerFactory testHostManagerFactory)
+        {
+            this.testHostManagerFactory = testHostManagerFactory;
+            this.executionManager = this.testHostManagerFactory.GetExecutionManager();
+        }
+
+        /// <summary>
+        /// Initialize adapters.
+        /// </summary>
+        public void Initialize()
+        {
+            if (!this.IsInitialized)
+            {
+                // We don't need to pass list of extension as we are running inside vstest.console and
+                // it will use TestPluginCache of vstest.console
+                executionManager.Initialize(Enumerable.Empty<string>());
+                this.IsInitialized = true;
+            }
+        }
+
+        /// <inheritdoc/>
+        public int StartTestRun(TestRunCriteria testRunCriteria, ITestRunEventsHandler eventHandler)
+        {
+            try
+            {
+                // Initialize extension before execution if not already initialized
+                this.Initialize();
+
+                var executionContext = new TestExecutionContext(
+                            testRunCriteria.FrequencyOfRunStatsChangeEvent,
+                            testRunCriteria.RunStatsChangeEventTimeout,
+                            inIsolation: false,
+                            keepAlive: testRunCriteria.KeepAlive,
+                            isDataCollectionEnabled: false,
+                            areTestCaseLevelEventsRequired: false,
+                            hasTestRun: true,
+                            isDebug: (testRunCriteria.TestHostLauncher != null && testRunCriteria.TestHostLauncher.IsDebug),
+                            testCaseFilter: testRunCriteria.TestCaseFilter);
+
```

Comment:
We should ask TestRuntimeProvider for extensions.

Reply:
Yes, we should.